pallets/xcm-rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,256 @@
+
+		/// Convert from `MultiLocation` to local `AssetId`
+		type CurrencyIdConvert: Convert<MultiLocation, Option<Self::AssetId>>;
+
+		/// Xcm rate limit getter for each asset
+		type RateLimitFor: GetByKey<Self::AssetId, Option<u128>>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	/// Accumulated amounts for each asset
+	#[pallet::getter(fn accumulated_amount)]
+	pub type AccumulatedAmounts<T: Config> =
+		StorageMap<_, Blake2_128Concat, MultiLocation, AccumulatedAmount, ValueQuery>;
+
+	#[pallet::event]
+	pub enum Event<T: Config> {}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+fn get_loc_and_amount(m: &MultiAsset) -> Option<(MultiLocation, u128)> {
+	match m.id {
+		AssetId::Concrete(location) => match m.fun {
+			Fungibility::Fungible(amount) => Some((location, amount)),
+			_ => None,
+		},
+		_ => None,
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn get_locations_and_amounts(instruction: &Instruction<T::RuntimeCall>) -> Vec<(MultiLocation, u128)> {
+		use Instruction::*;
+		match instruction {
+			// NOTE: This does not address the native asset "coming back" from other chains.
+			ReserveAssetDeposited(multi_assets) | ReceiveTeleportedAsset(multi_assets) => {
+				multi_assets.inner().iter().flat_map(get_loc_and_amount).collect()
+			}
+			_ => Vec::new(),
+		}
+	}
+}
+
+impl<T: Config> XcmDeferFilter<T::RuntimeCall> for Pallet<T> {
+	fn deferred_by(
+		_para: polkadot_parachain::primitives::Id,
+		_sent_at: RelayChainBlockNumber,
+		versioned_xcm: &VersionedXcm<T::RuntimeCall>,
+	) -> (Weight, Option<RelayChainBlockNumber>) {
+		use xcm::IntoVersion;
+		let maybe_xcm = versioned_xcm.clone().into_version(3);
+		let Ok(V3(xcm)) = maybe_xcm else { return (Weight::default(), Some(T::MaxDeferDuration::get())) };
+		// SAFETY NOTE: It is fine to only look at the first instruction because that is how assets will arrive on chain.
+		//              This is guaranteed by `AllowTopLevelExecution` which is standard in the ecosystem.
+		let Some(instruction) = xcm.first() else { return (Weight::default(), None) };
+		let mut total_weight = Weight::default();
+		let mut total_deferred_by: RelayChainBlockNumber = 0;
+		for (location, amount) in Pallet::<T>::get_locations_and_amounts(instruction) {
+			let accumulated_liquidity = AccumulatedAmounts::<T>::get(location);
+
+			// We assume that it's fine to not track assets whose id cannot be determined...
+			let Some(asset_id) = T::CurrencyIdConvert::convert(location) else {
+				total_weight.saturating_accrue(T::DbWeight::get().reads(1));
+				continue
+			};
+			// ... or that don't have a rate limit configured.
+			let Some(limit_per_duration) = T::RateLimitFor::get(&asset_id) else {
+				total_weight.saturating_accrue(T::DbWeight::get().reads(2));
```

Comment:
I assume the 1 one the read here refers to RateLimirFor::get().

Have you considered that eg T::RateLimitFor::get() can do more than 1 read?   

It is generic , so provided implementation might do even more reads.

Reply:
Yes, we considered it and decided against doing the benchmarking to ship faster.
Happy to document it more obviously if you want.